Maintainerr can't create jellyfin collections

**Describe the bug**
Not sure if it's related to https://github.com/Maintainerr/Maintainerr/issues/2489 ... it's very technical and on my side it seems like it does everything but create the collections

**To Reproduce**
Steps to reproduce the behavior:
1. Setup maintainerr with jellyfin
2. Test the rule, see if the rules get anything?
3. Check the collections created (they are empty)

**Expected behavior**
Get the collections on jellyfin and the collections on maintainerr should start to say they have media.

**Screenshots**

<img width="2356" height="582" alt="Image" src="https://github.com/user-attachments/assets/b3d961c3-f854-4062-ba0b-6111b4025637" />
Very simple .. the "test" one is literally this:

<img width="1880" height="872" alt="Image" src="https://github.com/user-attachments/assets/a787fcc9-2e72-4b30-ad36-a6d411a58684" />

and all media is caught:

<img width="1532" height="1410" alt="Image" src="https://github.com/user-attachments/assets/2f903c78-3a10-49cc-ac48-58c1c9f33c3d" />

**Device:**
 - OS: Ubuntu server (running it on the latest version on docker)
 - Version 3.1.0

**Additional context**
Here's the logs: 
```
[maintainerr] | 15/03/2026 08:00:11  [DEBUG] [CollectionsService] [checkAutomaticMediaServerLink] Setting mediaServerId to null — collection was empty or not found on media server
[maintainerr] | 15/03/2026 08:00:11  [ERROR] [JellyfinAdapterService] Failed to create Jellyfin collection
Error: Failed to fetch created collection
    at JellyfinAdapterService.createCollection (/opt/app/apps/server/dist/modules/api/media-server/jellyfin/jellyfin-adapter.service.js:629:23)
    at process.processTicksAndRejections (node:internal/process/task_queues:104:5)
    at async CollectionsService.addToCollection (/opt/app/apps/server/dist/modules/collections/collections.service.js:513:35)
    at async RuleExecutorService.handleCollection (/opt/app/apps/server/dist/modules/rules/tasks/rule-executor.service.js:304:30)
    at async RuleExecutorService.executeForRuleGroups (/opt/app/apps/server/dist/modules/rules/tasks/rule-executor.service.js:115:45)
    at async RuleExecutorJobManagerService.executeJob (/opt/app/apps/server/dist/modules/rules/tasks/rule-executor-job-manager.service.js:135:13)
    at async /opt/app/apps/server/dist/modules/rules/tasks/rule-executor-job-manager.service.js:118:21
[maintainerr] | 15/03/2026 08:00:11  [WARN] [CollectionsService] An error occurred while performing collection actions.
[maintainerr] | 15/03/2026 08:00:11  [WARN] [RuleExecutorService] Exception occurred while handling rule: Cannot read properties of undefined (reading 'id')
[maintainerr] | 15/03/2026 08:00:11  [INFO] [RuleExecutorService] Execution of rules for 'Sonarr to be removed' done.
[maintainerr] | 15/03/2026 08:00:11  [INFO] [RuleExecutorService] Starting execution of rule 'Radarr to be removed'
[maintainerr] | 15/03/2026 08:00:11  [INFO] [JellyfinAdapterService] Jellyfin connection test successful: Ody (10.11.6)
[maintainerr] | 15/03/2026 08:00:11  [INFO] [RuleExecutorService] Executing rules for 'Radarr to be removed'
[maintainerr] | 15/03/2026 08:00:11  [DEBUG] [CollectionsService] [checkAutomaticMediaServerLink] Collection "Radarr to be removed" (DB id: 3, mediaServerId: null)
[maintainerr] | 15/03/2026 08:00:11  [DEBUG] [CollectionsService] [checkAutomaticMediaServerLink] Setting mediaServerId to null — collection was empty or not found on media server
[maintainerr] | 15/03/2026 08:00:11  [DEBUG] [CollectionsService] [checkAutomaticMediaServerLink] Collection "Radarr to be removed" (DB id: 3, mediaServerId: null)
[maintainerr] | 15/03/2026 08:00:11  [DEBUG] [CollectionsService] [checkAutomaticMediaServerLink] Setting mediaServerId to null — collection was empty or not found on media server
[maintainerr] | 15/03/2026 08:00:11  [INFO] [RuleExecutorService] Execution of rules for 'Radarr to be removed' done.
[maintainerr] | 15/03/2026 08:00:11  [INFO] [RuleExecutorService] Starting execution of rule 'Test....'
[maintainerr] | 15/03/2026 08:00:11  [INFO] [JellyfinAdapterService] Jellyfin connection test successful: Ody (10.11.6)
[maintainerr] | 15/03/2026 08:00:12  [INFO] [RuleExecutorService] Executing rules for 'Test....'
[maintainerr] | 15/03/2026 08:00:15  [INFO] [RuleExecutorService] Adding 142 media items to 'Test....'.
[maintainerr] | 15/03/2026 08:00:15  [DEBUG] [CollectionsService] [checkAutomaticMediaServerLink] Collection "Test...." (DB id: 4, mediaServerId: null)
[maintainerr] | 15/03/2026 08:00:15  [DEBUG] [CollectionsService] [checkAutomaticMediaServerLink] Setting mediaServerId to null — collection was empty or not found on media server
[maintainerr] | 15/03/2026 08:00:15  [ERROR] [JellyfinAdapterService] Failed to create Jellyfin collection
Error: Failed to fetch created collection
    at JellyfinAdapterService.createCollection (/opt/app/apps/server/dist/modules/api/media-server/jellyfin/jellyfin-adapter.service.js:629:23)
    at process.processTicksAndRejections (node:internal/process/task_queues:104:5)
    at async CollectionsService.addToCollection (/opt/app/apps/server/dist/modules/collections/collections.service.js:513:35)
    at async RuleExecutorService.handleCollection (/opt/app/apps/server/dist/modules/rules/tasks/rule-executor.service.js:304:30)
    at async RuleExecutorService.executeForRuleGroups (/opt/app/apps/server/dist/modules/rules/tasks/rule-executor.service.js:115:45)
    at async RuleExecutorJobManagerService.executeJob (/opt/app/apps/server/dist/modules/rules/tasks/rule-executor-job-manager.service.js:135:13)
    at async /opt/app/apps/server/dist/modules/rules/tasks/rule-executor-job-manager.service.js:118:21
[maintainerr] | 15/03/2026 08:00:15  [WARN] [CollectionsService] An error occurred while performing collection actions.
[maintainerr] | 15/03/2026 08:00:15  [WARN] [RuleExecutorService] Exception occurred while handling rule: Cannot read properties of undefined (reading 'id')
[maintainerr] | 15/03/2026 08:00:15  [INFO] [RuleExecutorService] Execution of rules for 'Test....' done.
[maintainerr] | 15/03/2026 12:00:00  [INFO] [JellyfinAdapterService] Jellyfin connection test successful: Ody (10.11.6)
[maintainerr] | 15/03/2026 12:00:00  [INFO] [CollectionWorkerService] Started handling of all collections
[maintainerr] | 15/03/2026 12:00:00  [INFO] [CollectionWorkerService] Skipping collection 'Test....' as its action is 'Do Nothing'
[maintainerr] | 15/03/2026 12:00:00  [INFO] [CollectionWorkerService] Handling collection 'Sonarr to be removed'
[maintainerr] | 15/03/2026 12:00:00  [INFO] [CollectionWorkerService] Handling collection 'Sonarr to be removed' finished
[maintainerr] | 15/03/2026 12:00:00  [INFO] [CollectionWorkerService] Handling collection 'Radarr to be removed'
[maintainerr] | 15/03/2026 12:00:00  [INFO] [CollectionWorkerService] Handling collection 'Radarr to be removed' finished
[maintainerr] | 15/03/2026 12:00:00  [INFO] [CollectionWorkerService] All collections handled. No data was altered
[maintainerr] | 15/03/2026 12:00:00  [INFO] [CollectionWorkerService] Updating collection size cache...
[maintainerr] | 15/03/2026 12:00:00  [INFO] [CollectionWorkerService] Collection size cache updated
```

--- a/apps/server/src/modules/api/media-server/jellyfin/jellyfin-adapter.service.ts
+++ b/apps/server/src/modules/api/media-server/jellyfin/jellyfin-adapter.service.ts
@@ -17,6 +17,7 @@ import {
   getSystemApi,
   getTvShowsApi,
   getUserApi,
+  getUserLibraryApi,
 } from '@jellyfin/sdk/lib/utils/api/index.js';
 import {
   MediaServerFeature,
@@ -886,18 +887,14 @@ export class JellyfinAdapterService implements IMediaServerService {
 
     try {
       const userId = await this.getUserId();
-      const response = await getItemsApi(this.api).getItems({
+      const response = await getUserLibraryApi(this.api).getItem({
+        itemId: collectionId,
         userId,
-        ids: [collectionId],
-        fields: [
-          ItemFields.Overview,
-          ItemFields.DateCreated,
-          ItemFields.ChildCount,
-        ],
       });
 
-      const item = response.data.Items?.[0];
-      return item ? JellyfinMapper.toMediaCollection(item) : undefined;
+      return response.data
+        ? JellyfinMapper.toMediaCollection(response.data)
+        : undefined;
     } catch (error) {
       this.logger.warn(`Failed to get collection ${collectionId}`, error);
       return undefined;
@@ -928,12 +925,16 @@ export class JellyfinAdapterService implements IMediaServerService {
       // Note: No refresh needed - Jellyfin auto-generates composite images
       // when items are added (as long as isLocked: true, which we set above).
 
-      const collection = await this.getCollection(collectionId);
-      if (!collection) {
-        throw new Error('Failed to fetch created collection');
-      }
-
-      return collection;
+      // Construct from known data - the collection may not be immediately
+      // queryable via getItems as Jellyfin needs time to index it
+      return {
+        id: collectionId,
+        title: params.title,
+        summary: params.summary,
+        childCount: 0,
+        smart: false,
+        libraryId: params.libraryId,
+      };
     } catch (error) {
       this.logger.error('Failed to create Jellyfin collection', error);
       throw error;

--- a/apps/server/src/modules/collections/collections.service.ts
+++ b/apps/server/src/modules/collections/collections.service.ts
@@ -771,6 +771,23 @@ export class CollectionsService {
     media: AddRemoveCollectionMedia[],
     manual = false,
   ): Promise<Collection> {
+    return this.addToCollectionInternal(collectionDbId, media, manual, false);
+  }
+
+  async addToCollectionWithResolvedLink(
+    collection: Collection,
+    media: AddRemoveCollectionMedia[],
+    manual = false,
+  ): Promise<Collection> {
+    return this.addToCollectionInternal(collection.id, media, manual, true);
+  }
+
+  private async addToCollectionInternal(
+    collectionDbId: number,
+    media: AddRemoveCollectionMedia[],
+    manual = false,
+    skipAutomaticLinkCheck = false,
+  ): Promise<Collection> {
     try {
       const mediaServer = await this.getMediaServer();
       let collection = await this.collectionRepo.findOne({
@@ -787,7 +804,9 @@ export class CollectionsService {
       );
 
       if (collection) {
-        collection = await this.checkAutomaticMediaServerLink(collection);
+        if (!skipAutomaticLinkCheck) {
+          collection = await this.checkAutomaticMediaServerLink(collection);
+        }
 
         // Check if we need to create a new media server collection
         // This happens when: 1) we have new items to add, OR 2) we have existing items but no media server collection
@@ -908,6 +927,21 @@ export class CollectionsService {
     collectionDbId: number,
     media: AddRemoveCollectionMedia[],
   ) {
+    return this.removeFromCollectionInternal(collectionDbId, media, false);
+  }
+
+  async removeFromCollectionWithResolvedLink(
+    collection: Collection,
+    media: AddRemoveCollectionMedia[],
+  ) {
+    return this.removeFromCollectionInternal(collection.id, media, true);
+  }
+
+  private async removeFromCollectionInternal(
+    collectionDbId: number,
+    media: AddRemoveCollectionMedia[],
+    skipAutomaticLinkCheck = false,
+  ) {
     try {
       const mediaServer = await this.getMediaServer();
       let collection = await this.collectionRepo.findOne({
@@ -921,7 +955,10 @@ export class CollectionsService {
         return undefined;
       }
 
-      collection = await this.checkAutomaticMediaServerLink(collection);
+      if (!skipAutomaticLinkCheck) {
+        collection = await this.checkAutomaticMediaServerLink(collection);
+      }
+
       let collectionMedia = await this.CollectionMediaRepo.find({
         where: {
           collectionId: collectionDbId,

--- a/apps/server/src/modules/rules/tasks/rule-executor.service.spec.ts
+++ b/apps/server/src/modules/rules/tasks/rule-executor.service.spec.ts
@@ -48,7 +48,20 @@ describe('RuleExecutorService', () => {
           addedCount: items?.length ?? 0,
         };
       }),
+      addToCollectionWithResolvedLink: jest
+        .fn()
+        .mockImplementation(async (_collection, items) => {
+          return {
+            id: 1,
+            mediaServerId: 'coll-1',
+            title: 'Test Collection',
+            addedCount: items?.length ?? 0,
+          };
+        }),
       removeFromCollection: jest.fn().mockResolvedValue(undefined),
+      removeFromCollectionWithResolvedLink: jest
+        .fn()
+        .mockResolvedValue(undefined),
       saveCollection: jest.fn().mockResolvedValue(undefined),
       checkAutomaticMediaServerLink: jest
         .fn()
@@ -230,6 +243,136 @@ describe('RuleExecutorService', () => {
     expect(collectionService.addToCollection).not.toHaveBeenCalled();
   });
 
+  it('does not re-run addToCollection with no additions after resyncing a stale link', async () => {
+    const { service, collectionService } = createService(
+      MediaServerType.JELLYFIN,
+    );
+
+    const staleCollection = {
+      id: 1,
+      title: 'Test Collection',
+      mediaServerId: 'stale-coll',
+      manualCollection: false,
+      deleteAfterDays: 0,
+    };
+
+    collectionService.getCollection.mockResolvedValue(staleCollection as any);
+    collectionService.getCollectionMedia.mockResolvedValue([
+      { mediaServerId: 'm1' },
+    ] as any);
+    collectionService.checkAutomaticMediaServerLink.mockResolvedValueOnce({
+      ...staleCollection,
+      mediaServerId: null,
+    } as any);
+    collectionService.addToCollection.mockResolvedValueOnce({
+      ...staleCollection,
+      mediaServerId: 'new-coll',
+    } as any);
+    collectionService.saveCollection.mockImplementation(async (collection) => {
+      return collection as any;
+    });
+    collectionService.removeFromCollection.mockResolvedValue({
+      ...staleCollection,
+      mediaServerId: 'new-coll',
+    } as any);
+
+    (service as any).startTime = new Date();
+    (service as any).resultData = [];
+    (service as any).statisticsData = [];
+
+    await (service as any).handleCollection({ id: 10, collectionId: 1 });
+
+    expect(collectionService.addToCollection).toHaveBeenCalledTimes(1);
+    expect(
+      collectionService.addToCollectionWithResolvedLink,
+    ).not.toHaveBeenCalled();
+    expect(collectionService.addToCollection).toHaveBeenCalledWith(
+      1,
+      [{ mediaServerId: 'm1' }],
+      false,
+    );
+    expect(
+      collectionService.removeFromCollectionWithResolvedLink,
+    ).toHaveBeenCalledWith(
+      expect.objectContaining({
+        id: 1,
+        mediaServerId: 'new-coll',
+      }),
+      [
+        {
+          mediaServerId: 'm1',
+          reason: {
+            type: 'media_removed_by_rule',
+            data: undefined,
+          },
+        },
+      ],
+    );
+  });
+
+  it('reuses the reconciled collection link for additions later in the same run', async () => {
+    const { service, collectionService } = createService(
+      MediaServerType.JELLYFIN,
+    );
+
+    const staleCollection = {
+      id: 1,
+      title: 'Test Collection',
+      mediaServerId: 'stale-coll',
+      manualCollection: false,
+      deleteAfterDays: 0,
+    };
+
+    collectionService.getCollection.mockResolvedValue(staleCollection as any);
+    collectionService.getCollectionMedia.mockResolvedValue([
+      { mediaServerId: 'm1' },
+    ] as any);
+    collectionService.checkAutomaticMediaServerLink.mockResolvedValueOnce({
+      ...staleCollection,
+      mediaServerId: null,
+    } as any);
+    collectionService.addToCollection
+      .mockResolvedValueOnce({
+        ...staleCollection,
+        mediaServerId: 'new-coll',
+      } as any)
+      .mockResolvedValueOnce({
+        ...staleCollection,
+        mediaServerId: 'new-coll',
+      } as any);
+    collectionService.saveCollection.mockImplementation(async (collection) => {
+      return collection as any;
+    });
+    collectionService.removeFromCollection.mockResolvedValue({
+      ...staleCollection,
+      mediaServerId: 'new-coll',
+    } as any);
+
+    (service as any).startTime = new Date();
+    (service as any).resultData = [{ id: 'm2' }];
+    (service as any).statisticsData = [];
+
+    await (service as any).handleCollection({ id: 10, collectionId: 1 });
+
+    expect(
+      collectionService.addToCollectionWithResolvedLink,
+    ).toHaveBeenCalledWith(
+      expect.objectContaining({
+        id: 1,
+        mediaServerId: 'new-coll',
+      }),
+      [
+        {
+          mediaServerId: 'm2',
+          reason: {
+            type: 'media_added_by_rule',
+            data: undefined,
+          },
+        },
+      ],
+    );
+  });
+
   it('emits failed and skips execution when rule group has no library assigned', async () => {
     const { service, rulesService, eventEmitter, progressManager } =
       createService(MediaServerType.JELLYFIN);
@@ -258,7 +401,7 @@ describe('RuleExecutorService', () => {
     (service as any).resultData = [{ id: 'm2' }];
     (service as any).statisticsData = [];
 
-    collectionService.addToCollection.mockImplementation(async () => {
+    const abortMock = async () => {
       abortController.abort();
       return {
         id: 1,
@@ -266,7 +409,11 @@ describe('RuleExecutorService', () => {
         title: 'Test Collection',
         deleteAfterDays: 0,
       } as any;
-    });
+    };
+    collectionService.addToCollection.mockImplementation(abortMock);
+    collectionService.addToCollectionWithResolvedLink.mockImplementation(
+      abortMock,
+    );
 
     await expect(
       (
@@ -280,5 +427,8 @@ describe('RuleExecutorService', () => {
     ).rejects.toMatchObject({ name: 'AbortError' });
 
     expect(collectionService.removeFromCollection).not.toHaveBeenCalled();
+    expect(
+      collectionService.removeFromCollectionWithResolvedLink,
+    ).not.toHaveBeenCalled();
   });
 });

--- a/apps/server/src/modules/rules/tasks/rule-executor.service.ts
+++ b/apps/server/src/modules/rules/tasks/rule-executor.service.ts
@@ -499,16 +499,32 @@ export class RuleExecutorService {
           await this.collectionService.relinkManualCollection(collection);
 
         abortSignal?.throwIfAborted();
-        collection = await this.collectionService.addToCollection(
-          collection.id,
-          dataToAdd,
-        );
+        if (dataToAdd.length > 0) {
+          collection =
+            collMediaData.length > 0
+              ? await this.collectionService.addToCollectionWithResolvedLink(
+                  collection,
+                  dataToAdd,
+                )
+              : await this.collectionService.addToCollection(
+                  collection.id,
+                  dataToAdd,
+                );
+        }
 
         abortSignal?.throwIfAborted();
-        collection = await this.collectionService.removeFromCollection(
-          collection.id,
-          dataToRemove,
-        );
+        if (dataToRemove.length > 0) {
+          collection =
+            collMediaData.length > 0
+              ? await this.collectionService.removeFromCollectionWithResolvedLink(
+                  collection,
+                  dataToRemove,
+                )
+              : await this.collectionService.removeFromCollection(
+                  collection.id,
+                  dataToRemove,
+                );
+        }
 
         // Determine which items were actually added/removed by comparing DB state
         const updatedMediaServerIds = new Set(

